docs/Operators.md

```diff
@@ -5410,6 +5454,51 @@ opset_import {
 </dl>
 
 
+### <sub>experimental</sub> <a name="LoopIndexTensor"></a><a name="loopindextensor">**LoopIndexTensor**</a>
+
```

Comment:
I suggest not tying this operator so tightly with a loop. Indexing a tensor is a natural operation, independent of loops. Let's call this just "IndexTensor" and not restrict it to inside loops.
This requires rationalizing the use of "loop_var" attribute as well. I think "loop_var" is a variable representing a dynamic runtime value, and so it is best represented as an input instead of as an attribute.

Reply:
@zdevito what do you think here?

Reply:
I think we should be very careful about an operator like this. Previously, all of our operators were pure operations: it doesn't matter when you execute them, you can inline them into subgraphs unconditionally. `LoopIndexTensor`, on the other hand, is sensitive to the particular graph from which it is run; you can't push it into a subgraph; there is a kind of dynamic scoping going on. Maybe we will find this class of operators sufficiently useful that we will enshrine them as a class of their own but I would take a close look at it.

Reply:
I agree with @ezyang 's point, but would say the issue is not with the operator itself, but the use of "loop_var" ... one possibility is to view it as a mutable variable (which don't exist in the current ONNX) or as shorthand: the loop-body is implicitly a function (lambda) with the loop_var being an implicit parameter.

Reply:
It would be far better to model this as dynamic scoping (which now has the correct properties), but you can always desugar dynamic scoping into explicit parameter passing and that keeps ONNX simpler.

Reply:
This PR raises some interesting questions around scoping of names within subgraphs. It can certainly be made to work, in several ways, but we would need to very carefully specify the scoping rules within hierarchies of graphs so that we can avoid implementation divergence.

From a name scoping perspective, it may be easier to introduce function definitions and define control flow operators in terms of function attributes rather than graphs. It would require explicitly passing everything you need into the subgraph rather than capturing them from the parent graph(s).

Just a thought,

Niklas